File: tests/algorithms/x/Java/graphs/graph_adjacency_list.java
public class Main {
    static class Graph {
        java.util.Map<String,String[]> adj;
        boolean directed;
        Graph(java.util.Map<String,String[]> adj, boolean directed) {
            this.adj = adj;
            this.directed = directed;
        }
        Graph() {}
        @Override public String toString() {
            return String.format("{'adj': %s, 'directed': %s}", String.valueOf(adj), String.valueOf(directed));
        }
    }


    static Graph create_graph(String[] vertices, String[][] edges, boolean directed) {
        java.util.Map<String,String[]> adj = ((java.util.Map<String,String[]>)(new java.util.LinkedHashMap<String, String[]>()));
        for (String v : vertices) {
adj.put(v, ((String[])(new String[]{})));
        }
        for (String[] e : edges) {
            String s = e[0];
            String d = e[1];
            if (!(Boolean)(adj.containsKey(s))) {
adj.put(s, ((String[])(new String[]{})));
            }
            if (!(Boolean)(adj.containsKey(d))) {
adj.put(d, ((String[])(new String[]{})));
            }
adj.put(s, ((String[])(java.util.stream.Stream.concat(java.util.Arrays.stream(((String[])(adj).get(s))), java.util.stream.Stream.of(d)).toArray(String[]::new))));
            if (!(Boolean)directed) {
adj.put(d, ((String[])(java.util.stream.Stream.concat(java.util.Arrays.stream(((String[])(adj).get(d))), java.util.stream.Stream.of(s)).toArray(String[]::new))));
            }
        }
        return new Graph(adj, directed);
    }

    static Graph add_vertex(Graph graph, String v) {
        if (((Boolean)(graph.adj.containsKey(v)))) {
            throw new RuntimeException(String.valueOf("vertex exists"));
        }
        java.util.Map<String,String[]> adj_1 = ((java.util.Map<String,String[]>)(new java.util.LinkedHashMap<String, String[]>()));
        for (String k : graph.adj.keySet()) {
adj_1.put(k, (String[])(((String[])(graph.adj).get(k))));
        }
adj_1.put(v, ((String[])(new String[]{})));
        return new Graph(adj_1, graph.directed);
    }

    static String[] remove_from_list(String[] lst, String value) {
        String[] res = ((String[])(new String[]{}));
        int i = 0;
        while (i < lst.length) {
            if (!(lst[i].equals(value))) {
                res = ((String[])(java.util.stream.Stream.concat(java.util.Arrays.stream(res), java.util.stream.Stream.of(lst[i])).toArray(String[]::new)));
            }
            i = i + 1;
        }
        return res;
    }

    static java.util.Map<String,String[]> remove_key(java.util.Map<String,String[]> m, String key) {
        java.util.Map<String,String[]> res_1 = ((java.util.Map<String,String[]>)(new java.util.LinkedHashMap<String, String[]>()));
        for (String k : m.keySet()) {
            if (((Number)(k)).intValue() != key) {
res_1.put(k, (String[])(((String[])(m).get(k))));
            }
        }
        return res_1;
    }

    static Graph add_edge(Graph graph, String s, String d) {
        if (((!(Boolean)(graph.adj.containsKey(s))) || (!(Boolean)(graph.adj.containsKey(d))))) {
            throw new RuntimeException(String.valueOf("vertex missing"));
        }
        if (((Boolean)(contains_edge(graph, s, d)))) {
            throw new RuntimeException(String.valueOf("edge exists"));
        }
        java.util.Map<String,String[]> adj_2 = ((java.util.Map<String,String[]>)(new java.util.LinkedHashMap<String, String[]>()));
        for (String k : graph.adj.keySet()) {
adj_2.put(k, (String[])(((String[])(graph.adj).get(k))));
        }
        String[] list_s = (String[])(((String[])(adj_2).get(s)));
        list_s = ((String[])(java.util.stream.Stream.concat(java.util.Arrays.stream(list_s), java.util.stream.Stream.of(d)).toArray(String[]::new)));
adj_2.put(s, ((String[])(list_s)));
        if (!graph.directed) {
            String[] list_d = (String[])(((String[])(adj_2).get(d)));
            list_d = ((String[])(java.util.stream.Stream.concat(java.util.Arrays.stream(list_d), java.util.stream.Stream.of(s)).toArray(String[]::new)));
adj_2.put(d, ((String[])(list_d)));
        }
        return new Graph(adj_2, graph.directed);
    }

    static Graph remove_edge(Graph graph, String s, String d) {
        if (((!(Boolean)(graph.adj.containsKey(s))) || (!(Boolean)(graph.adj.containsKey(d))))) {
            throw new RuntimeException(String.valueOf("vertex missing"));
        }
        if (!(Boolean)contains_edge(graph, s, d)) {
            throw new RuntimeException(String.valueOf("edge missing"));
        }
        java.util.Map<String,String[]> adj_3 = ((java.util.Map<String,String[]>)(new java.util.LinkedHashMap<String, String[]>()));
        for (String k : graph.adj.keySet()) {
adj_3.put(k, (String[])(((String[])(graph.adj).get(k))));
        }
adj_3.put(s, ((String[])(remove_from_list((String[])(((String[])(adj_3).get(s))), d))));
        if (!graph.directed) {
adj_3.put(d, ((String[])(remove_from_list((String[])(((String[])(adj_3).get(d))), s))));
        }
        return new Graph(adj_3, graph.directed);
    }

    static Graph remove_vertex(Graph graph, String v) {
        if (!(Boolean)(graph.adj.containsKey(v))) {
            throw new RuntimeException(String.valueOf("vertex missing"));
        }
        java.util.Map<String,String[]> adj_4 = ((java.util.Map<String,String[]>)(new java.util.LinkedHashMap<String, String[]>()));
        for (String k : graph.adj.keySet()) {
            if (((Number)(k)).intValue() != v) {
adj_4.put(k, ((String[])(remove_from_list((String[])(((String[])(graph.adj).get(k))), v))));
            }
        }
        return new Graph(adj_4, graph.directed);
    }

    static boolean contains_vertex(Graph graph, String v) {
        return graph.adj.containsKey(v);
    }

    static boolean contains_edge(Graph graph, String s, String d) {
        if (((!(Boolean)(graph.adj.containsKey(s))) || (!(Boolean)(graph.adj.containsKey(d))))) {
            throw new RuntimeException(String.valueOf("vertex missing"));
        }
        for (String x : ((String[])(graph.adj).get(s))) {
            if ((x.equals(d))) {
                return true;
            }
        }
        return false;
    }

    static Graph clear_graph(Graph graph) {
        return new Graph(new java.util.LinkedHashMap<String, String[]>(), graph.directed);
    }

    static String to_string(Graph graph) {
        return _p(graph.adj);
    }

    static void main() {
        String[] vertices = ((String[])(new String[]{"1", "2", "3", "4"}));
        String[][] edges = ((String[][])(new String[][]{new String[]{"1", "2"}, new String[]{"2", "3"}, new String[]{"3", "4"}}));
        Graph g = create_graph(((String[])(vertices)), ((String[][])(edges)), false);
        System.out.println(to_string(g));
        g = add_vertex(g, "5");
        g = add_edge(g, "4", "5");
        System.out.println(_p(contains_edge(g, "4", "5")));
        g = remove_edge(g, "1", "2");
        g = remove_vertex(g, "3");
        System.out.println(to_string(g));
    }
    public static void main(String[] args) {
        {
            long _benchStart = _now();
            long _benchMem = _mem();
            main();
            long _benchDuration = _now() - _benchStart;
            long _benchMemory = _mem() - _benchMem;
            System.out.println("{");
            System.out.println("  \"duration_us\": " + _benchDuration + ",");
            System.out.println("  \"memory_bytes\": " + _benchMemory + ",");
            System.out.println("  \"name\": \"main\"");
            System.out.println("}");
            return;
        }
    }

    static boolean _nowSeeded = false;
    static int _nowSeed;
    static int _now() {
        if (!_nowSeeded) {
            String s = System.getenv("MOCHI_NOW_SEED");
            if (s != null && !s.isEmpty()) {
                try { _nowSeed = Integer.parseInt(s); _nowSeeded = true; } catch (Exception e) {}
            }
        }
        if (_nowSeeded) {
            _nowSeed = (int)((_nowSeed * 1664525L + 1013904223) % 2147483647);
            return _nowSeed;
        }
        return (int)(System.nanoTime() / 1000);
    }

    static long _mem() {
        Runtime rt = Runtime.getRuntime();
        rt.gc();
        return rt.totalMemory() - rt.freeMemory();
    }

    static String _p(Object v) {
        if (v == null) return "<nil>";
        if (v.getClass().isArray()) {
            if (v instanceof int[]) return java.util.Arrays.toString((int[]) v);
            if (v instanceof long[]) return java.util.Arrays.toString((long[]) v);
            if (v instanceof double[]) return java.util.Arrays.toString((double[]) v);
            if (v instanceof boolean[]) return java.util.Arrays.toString((boolean[]) v);
            if (v instanceof byte[]) return java.util.Arrays.toString((byte[]) v);
            if (v instanceof char[]) return java.util.Arrays.toString((char[]) v);
            if (v instanceof short[]) return java.util.Arrays.toString((short[]) v);
            if (v instanceof float[]) return java.util.Arrays.toString((float[]) v);
            return java.util.Arrays.deepToString((Object[]) v);
        }
        return String.valueOf(v);
    }
}

File: tests/algorithms/transpiler/FS/graphs/graph_adjacency_list.fs
// Generated 2025-08-08 16:03 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _dictAdd<'K,'V when 'K : equality> (d:System.Collections.Generic.IDictionary<'K,'V>) (k:'K) (v:'V) =
    d.[k] <- v
    d
let _dictCreate<'K,'V when 'K : equality> (pairs:('K * 'V) list) : System.Collections.Generic.IDictionary<'K,'V> =
    let d = System.Collections.Generic.Dictionary<'K, 'V>()
    for (k, v) in pairs do
        d.[k] <- v
    upcast d
let _dictGet<'K,'V when 'K : equality> (d:System.Collections.Generic.IDictionary<'K,'V>) (k:'K) : 'V =
    match d.TryGetValue(k) with
    | true, v -> v
    | _ -> Unchecked.defaultof<'V>
let _idx (arr:'a array) (i:int) : 'a =
    if not (obj.ReferenceEquals(arr, null)) && i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let rec _str v =
    let s = sprintf "%A" v
    s.Replace("[|", "[")
     .Replace("|]", "]")
     .Replace("; ", " ")
     .Replace(";", "")
     .Replace("\"", "")
type Graph = {
    mutable _adj: System.Collections.Generic.IDictionary<string, string array>
    mutable _directed: bool
}
open System.Collections.Generic

let rec create_graph (vertices: string array) (edges: string array array) (_directed: bool) =
    let mutable __ret : Graph = Unchecked.defaultof<Graph>
    let mutable vertices = vertices
    let mutable edges = edges
    let mutable _directed = _directed
    try
        let mutable _adj: System.Collections.Generic.IDictionary<string, string array> = _dictCreate []
        for v in Seq.map string (vertices) do
            _adj.[v] <- [||]
        for e in edges do
            let s: string = _idx e (0)
            let d: string = _idx e (1)
            if not (_adj.ContainsKey(s)) then
                _adj.[s] <- [||]
            if not (_adj.ContainsKey(d)) then
                _adj.[d] <- [||]
            _adj.[s] <- Array.append (_dictGet _adj ((string (s)))) [|d|]
            if not _directed then
                _adj.[d] <- Array.append (_dictGet _adj ((string (d)))) [|s|]
        __ret <- { _adj = _adj; _directed = _directed }
        raise Return
        __ret
    with
        | Return -> __ret
and add_vertex (graph: Graph) (v: string) =
    let mutable __ret : Graph = Unchecked.defaultof<Graph>
    let mutable graph = graph
    let mutable v = v
    try
        if (graph._adj).ContainsKey(v) then
            failwith ("vertex exists")
        let mutable _adj: System.Collections.Generic.IDictionary<string, string array> = _dictCreate []
        for k in (graph._adj).Keys do
            _adj.[k] <- _dictGet (graph._adj) ((string (k)))
        _adj.[v] <- [||]
        __ret <- { _adj = _adj; _directed = graph._directed }
        raise Return
        __ret
    with
        | Return -> __ret
and remove_from_list (lst: string array) (value: string) =
    let mutable __ret : string array = Unchecked.defaultof<string array>
    let mutable lst = lst
    let mutable value = value
    try
        let mutable res: string array = [||]
        let mutable i: int = 0
        while i < (Seq.length (lst)) do
            if (_idx lst (i)) <> value then
                res <- Array.append res [|(_idx lst (i))|]
            i <- i + 1
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
and remove_key (m: System.Collections.Generic.IDictionary<string, string array>) (key: string) =
    let mutable __ret : System.Collections.Generic.IDictionary<string, string array> = Unchecked.defaultof<System.Collections.Generic.IDictionary<string, string array>>
    let mutable m = m
    let mutable key = key
    try
        let mutable res: System.Collections.Generic.IDictionary<string, string array> = _dictCreate []
        for k in m.Keys do
            if k <> key then
                res.[k] <- _dictGet m ((string (k)))
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
and add_edge (graph: Graph) (s: string) (d: string) =
    let mutable __ret : Graph = Unchecked.defaultof<Graph>
    let mutable graph = graph
    let mutable s = s
    let mutable d = d
    try
        if (not ((graph._adj).ContainsKey(s))) || (not ((graph._adj).ContainsKey(d))) then
            failwith ("vertex missing")
        if contains_edge (graph) (s) (d) then
            failwith ("edge exists")
        let mutable _adj: System.Collections.Generic.IDictionary<string, string array> = _dictCreate []
        for k in (graph._adj).Keys do
            _adj.[k] <- _dictGet (graph._adj) ((string (k)))
        let mutable list_s: string array = _dictGet _adj ((string (s)))
        list_s <- Array.append list_s [|d|]
        _adj.[s] <- list_s
        if not (graph._directed) then
            let mutable list_d: string array = _dictGet _adj ((string (d)))
            list_d <- Array.append list_d [|s|]
            _adj.[d] <- list_d
        __ret <- { _adj = _adj; _directed = graph._directed }
        raise Return
        __ret
    with
        | Return -> __ret
and remove_edge (graph: Graph) (s: string) (d: string) =
    let mutable __ret : Graph = Unchecked.defaultof<Graph>
    let mutable graph = graph
    let mutable s = s
    let mutable d = d
    try
        if (not ((graph._adj).ContainsKey(s))) || (not ((graph._adj).ContainsKey(d))) then
            failwith ("vertex missing")
        if not (contains_edge (graph) (s) (d)) then
            failwith ("edge missing")
        let mutable _adj: System.Collections.Generic.IDictionary<string, string array> = _dictCreate []
        for k in (graph._adj).Keys do
            _adj.[k] <- _dictGet (graph._adj) ((string (k)))
        _adj.[s] <- remove_from_list (_dictGet _adj ((string (s)))) (d)
        if not (graph._directed) then
            _adj.[d] <- remove_from_list (_dictGet _adj ((string (d)))) (s)
        __ret <- { _adj = _adj; _directed = graph._directed }
        raise Return
        __ret
    with
        | Return -> __ret
and remove_vertex (graph: Graph) (v: string) =
    let mutable __ret : Graph = Unchecked.defaultof<Graph>
    let mutable graph = graph
    let mutable v = v
    try
        if not ((graph._adj).ContainsKey(v)) then
            failwith ("vertex missing")
        let mutable _adj: System.Collections.Generic.IDictionary<string, string array> = _dictCreate []
        for k in (graph._adj).Keys do
            if k <> v then
                _adj.[k] <- remove_from_list (_dictGet (graph._adj) ((string (k)))) (v)
        __ret <- { _adj = _adj; _directed = graph._directed }
        raise Return
        __ret
    with
        | Return -> __ret
and contains_vertex (graph: Graph) (v: string) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable graph = graph
    let mutable v = v
    try
        __ret <- (graph._adj).ContainsKey(v)
        raise Return
        __ret
    with
        | Return -> __ret
and contains_edge (graph: Graph) (s: string) (d: string) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable graph = graph
    let mutable s = s
    let mutable d = d
    try
        if (not ((graph._adj).ContainsKey(s))) || (not ((graph._adj).ContainsKey(d))) then
            failwith ("vertex missing")
        for x in Seq.map string (_dictGet (graph._adj) ((string (s)))) do
            if x = d then
                __ret <- true
                raise Return
        __ret <- false
        raise Return
        __ret
    with
        | Return -> __ret
and clear_graph (graph: Graph) =
    let mutable __ret : Graph = Unchecked.defaultof<Graph>
    let mutable graph = graph
    try
        __ret <- { _adj = _dictCreate []; _directed = graph._directed }
        raise Return
        __ret
    with
        | Return -> __ret
and to_string (graph: Graph) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable graph = graph
    try
        __ret <- _str (graph._adj)
        raise Return
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        let __bench_start = _now()
        let __mem_start = System.GC.GetTotalMemory(true)
        let vertices: string array = [|"1"; "2"; "3"; "4"|]
        let edges: string array array = [|[|"1"; "2"|]; [|"2"; "3"|]; [|"3"; "4"|]|]
        let mutable g: Graph = create_graph (vertices) (edges) (false)
        printfn "%s" (to_string (g))
        g <- add_vertex (g) ("5")
        g <- add_edge (g) ("4") ("5")
        printfn "%s" (_str (contains_edge (g) ("4") ("5")))
        g <- remove_edge (g) ("1") ("2")
        g <- remove_vertex (g) ("3")
        printfn "%s" (to_string (g))
        let __bench_end = _now()
        let __mem_end = System.GC.GetTotalMemory(true)
        printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

        __ret
    with
        | Return -> __ret
main()
